components/style/values/specified/image.rs

```diff
@@ -0,0 +1,413 @@
+    fn to_css<W>(&self, dest: &mut W) -> fmt::Result where W: fmt::Write {
+        try!(self.color.to_css(dest));
+        if let Some(position) = self.position {
+            try!(dest.write_str(" "));
+            try!(position.to_css(dest));
+        }
+        Ok(())
+    }
+}
+
+define_css_keyword_enum!(HorizontalDirection: "left" => Left, "right" => Right);
+define_css_keyword_enum!(VerticalDirection: "top" => Top, "bottom" => Bottom);
+
+impl Parse for ColorStop {
+    fn parse(input: &mut Parser) -> Result<Self, ()> {
+        Ok(ColorStop {
+            color: try!(CSSColor::parse(input)),
+            position: input.try(LengthOrPercentage::parse).ok(),
+        })
+    }
+}
+
+/// Determines whether the gradient's ending shape is a circle or an ellipse.
+/// If <shape> is omitted, the ending shape defaults to a circle
+/// if the <size> is a single <length>, and to an ellipse otherwise.
+/// https://drafts.csswg.org/css-images/#valdef-radial-gradient-ending-shape
+#[derive(Clone, PartialEq, Debug)]
+#[cfg_attr(feature = "servo", derive(HeapSizeOf))]
+pub enum EndingShape {
+    Circle(LengthOrKeyword),
+    Ellipse(LengthOrPercentageOrKeyword, LengthOrPercentageOrKeyword),
+}
+
+impl Parse for EndingShape {
+    fn parse(input: &mut Parser) -> Result<Self, ()> {
+        match_ignore_ascii_case! { try!(input.expect_ident()),
+            "circle" => {
+                let position = input.try(LengthOrKeyword::parse).unwrap_or(
+                    LengthOrKeyword::Keyword(SizeKeyword::FarthestSide));
+                Ok(EndingShape::Circle(position))
+            },
+            "ellipse" => {
+              let (first_pos, second_pos) = if let Ok(first) = input.try(LengthOrPercentageOrKeyword::parse) {
+                    if let Ok(second) = input.try(LengthOrPercentageOrKeyword::parse) {
+                        (first, second)
+                    } else {
+                        (first, LengthOrPercentageOrKeyword::Keyword(SizeKeyword::FarthestSide))
+                    }
+                } else {
+                    (LengthOrPercentageOrKeyword::Keyword(SizeKeyword::FarthestSide),
+                     LengthOrPercentageOrKeyword::Keyword(SizeKeyword::FarthestSide))
+                };
+                Ok(EndingShape::Ellipse(first_pos, second_pos))
+            },
+            _ => {
+                // If two <length> is present, it defaults to circle, otherwise defaults to ellipse.
+                if let Ok(first) = input.try(LengthOrPercentageOrKeyword::parse) {
+                    if let Ok(second) = input.try(LengthOrPercentageOrKeyword::parse) {
+                        Ok(EndingShape::Ellipse(first, second))
+                    } else {
+                        if let LengthOrPercentageOrKeyword::LengthOrPercentage(length) = first {
+                            if let LengthOrPercentage::Length(len) = length {
+                                Ok(EndingShape::Circle(LengthOrKeyword::Length(len)))
```

Comment:
We're still copying over the length. Don't do this (because of calc and how lengths parse differently), try to parse both LoPoK first, and if they fail, try parsing an LoK then.